app/graphql/types/profile.rb

```diff
@@ -62,39 +48,99 @@ def rules(args = {})
 
+    def rules(args = {})
+      context[:parent_profile_id] ||= {}
+      context[:rule_results] ||= {} if args[:lookahead].selects?(:compliant)
+      if system_id(args)
+        latest_test_result_batch(args).then do |latest_test_result|
+          ::CollectionLoader.for(::TestResult, :rule_results).load(latest_test_result).then do |rule_results|
+            if args[:lookahead].selects?(:references)
+              grouped_rules_references = ::RuleReferencesRule.where(rule_id: rule_results.pluck(:rule_id)).group_by(&:rule_id)
+              grouped_rules_references.each do |rule_id, references|
+                context[:"rule_references_#{rule_id}"] = references.pluck(:rule_reference_id)
+              end
+            end
+            if rule_results.blank?
+              []
+            else
+              ::RecordLoader.for(::Rule).load_many(rule_results.pluck(:rule_id)).then do |rules|
+                rules.each do |rule|
+                  context[:parent_profile_id][rule.id] = object.id
+                end
+                rule_results.each do |rule_result|
+                  context[:rule_results][rule_result.rule_id] ||= {}
+                  context[:rule_results][rule_result.rule_id][object.id] = rule_result.result
+                end if args[:lookahead].selects?(:compliant)
+                rules
+              end
+            end
+          end
+        end
+      else
+        ::CollectionLoader.for(::Profile, :rules).load(object).then do |rules|
+          rules
+        end
+      end
+    end
+
     def compliant_host_count
-      object.hosts.count { |host| object.compliant?(host) }
+      CollectionLoader.for(object.class, :hosts).load(object).then do |hosts|
+        hosts.count { |host| object.compliant?(host) }
+      end
     end
 
     def total_host_count
-      object.hosts.count
+      CollectionLoader.for(object.class, :hosts).load(object).then(&:count)
     end
 
     def compliant(args = {})
-      object.compliant?(Host.find(system_id(args)))
+      latest_test_result_batch(args).then do |latest_test_result|
+        host_results = latest_test_result&.rule_results
+        host_results.present? &&
+          (
+            host_results.where(result: 'pass').count /
+            host_results.count.to_f
+          ) >=
+            (object.compliance_threshold / 100.0)
+      end
     end
 
     def rules_passed(args = {})
-      Host.find(system_id(args)).rules_passed(object)
+      latest_test_result_batch(args).then do |latest_test_result|
+        latest_test_result
+          &.rule_results
+          &.where(result: %w[pass notapplicable notselected])&.count
+      end
     end
 
     def rules_failed(args = {})
-      Host.find(system_id(args)).rules_failed(object)
+      latest_test_result_batch(args).then do |latest_test_result|
+        latest_test_result
+          &.rule_results
+          &.where&.not(result: %w[pass notapplicable notselected])&.count
```

Comment:
All these safe operators, but all these methods are guaranteed to return an AR::Relation and never `nil`.